### PR TITLE
ddl: generated columns to be disallowed more completely for `column type change` operation (need to change the data)

### LINKED ISSUE #24310: Modifying the column type(need to change the data) involved in the generated column is not completely prohibited
## Bug Report

Please answer these questions before submitting your issue. Thanks!

There is currently no support for changing the type of a column involving generating a column. Related to [compatibility-issues-with-mysql](https://github.com/pingcap/tidb/blob/master/docs/design/2020-07-07-change-column-types.md#compatibility-issues-with-mysql).

### 1. Minimal reproduce step (Required)
```
 create table t2(id int, a int, b int generated always as (abs(a)) virtual);
 alter table t2 modify column a mediumint;
```

### 2. What did you expect to see? (Required)
Return an error like `Unsupported modify column: tidb_enable_change_column_type is true, oldCol is a dependent column 'a' for generated column`.

### 3. What did you see instead (Required)
`Query OK, 0 rows affected (2.52 sec)`

### 4. What is your TiDB version? (Required)
```
Release Version: v4.0.0-beta.2-2732-gd9e2bd3da
Edition: Community
Git Commit Hash: d9e2bd3da49ec3e9a062831fb6e4ab1219a8d158
Git Branch: master
UTC Build Time: 2021-04-27 06:27:55
```

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close #24310 

Problem Summary:
Modifying the column type(need to change the data) involved in the generated column is not completely prohibited。

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Checks that the column that modifies the column type is a dependent column of a generated column.

### Check List 

Tests 

- Unit test

### Release note 

- Changing column types involved in generated columns are completely disallowed (data needs to be changed)

### CODE REVIEW

File: ddl/generated_column.go [hunk 1]
```diff
@@ -137,6 +139,31 @@ func findColumnNamesInExpr(expr ast.ExprNode) []*ast.ColumnName {
...
+			}
+		}
+	}
+	return false, ""
+}
+
+func isGeneratedRelatedColumn(tblInfo *model.TableInfo, newCol, col *model.ColumnInfo) error {
+	if newCol.IsGenerated() || col.IsGenerated() {
+		// TODO: Make it compatible with MySQL error.
+		msg := fmt.Sprintf("tidb_enable_change_column_type is true, newCol IsGenerated %v, oldCol IsGenerated %v", newCol.IsGenerated(), col.IsGenerated())
```
Comment (AilinKid) [inline ddl/generated_column.go:157]: `IsGenerated` is less friendly to read in the log.

**Review by AilinKid — COMMENTED** [review]
rest lgtm

File: ddl/generated_column.go (hunk 1 above)
Comment (zimulala) [reply ddl/generated_column.go:157]: Any suggestions?

Comment (AilinKid) [reply ddl/generated_column.go:157]: it not big deal, i think

**Review by tangenta — APPROVED** [review]

**Pull request merged**